src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserReleaseRoleRepository.cs

```diff
@@ -20,20 +21,94 @@ public UserReleaseRoleRepository(ContentDbContext contentDbContext)
+                Role = role,
+                Created = DateTime.UtcNow,
+                CreatedById = createdById,
             };
 
-            var created = (await _contentDbContext.UserReleaseRoles.AddAsync(userReleaseRole)).Entity;
+            var created =
+                (await _contentDbContext.UserReleaseRoles.AddAsync(userReleaseRole)).Entity;
             await _contentDbContext.SaveChangesAsync();
             return created;
         }
 
+        public async Task<Unit> CreateAll(List<Guid> userIds, Guid releaseId, ReleaseRole role,
+            Guid createdById)
+        {
+            var userIdsAlreadyHaveRole = await _contentDbContext.UserReleaseRoles
+                .AsAsyncEnumerable()
+                .Where(urr =>
+                    urr.ReleaseId == releaseId
+                    && urr.Role == role
+                    && userIds.Contains(urr.UserId))
+                .Select(urr => urr.UserId)
+                .ToListAsync();
+
+            var userIdsToAdd = userIds.Except(userIdsAlreadyHaveRole).ToList();
+
+            var newUserReleaseRoles = userIdsToAdd
+                .Select(userId =>
+                    new UserReleaseRole
+                    {
+                        UserId = userId,
+                        ReleaseId = releaseId,
+                        Role = role,
+                        Created = DateTime.UtcNow,
+                        CreatedById = createdById,
+                    }
+                ).ToList();
+
+            await _contentDbContext.UserReleaseRoles.AddRangeAsync(newUserReleaseRoles);
+            await _contentDbContext.SaveChangesAsync();
+            return Unit.Instance;
+        }
+
+        public async Task Remove(UserReleaseRole userReleaseRole, Guid deletedById)
+        {
+            userReleaseRole.Deleted = DateTime.UtcNow;
+            userReleaseRole.DeletedById = deletedById;
+            _contentDbContext.Update(userReleaseRole);
+            await _contentDbContext.SaveChangesAsync();
+        }
+
+        public async Task RemoveAll(List<UserReleaseRole> userReleaseRoles, Guid deletedById)
+        {
+            userReleaseRoles.ForEach(userReleaseRole =>
+            {
+                userReleaseRole.Deleted = DateTime.UtcNow;
+                userReleaseRole.DeletedById = deletedById;
+            });
+            _contentDbContext.UpdateRange(userReleaseRoles);
+            await _contentDbContext.SaveChangesAsync();
+        }
+
+        public async Task RemoveAllUserReleaseRolesForPublication(Guid userId, Publication publication, ReleaseRole role, Guid deletedById)
+        {
+            _contentDbContext.Update(publication);
+            await _contentDbContext
+                .Entry(publication)
+                .Collection(p => p.Releases)
+                .LoadAsync();
+            var allLatestReleaseVersionIds = publication
+                .GetLatestVersionsOfAllReleases()
+                .Select(r => r.Id)
+                .ToList();
```

Comment:
There is an issue with permissions currently whereby you could create an amendment, remove a permission, and then cancel the amendment, and I believe the permission would be restored. Permissions are currently copied when you create an amendment.

My personal feeling is to leave this, as the highlighted code is currently consistent with how we update permissions elsewhere. If we decided to change this to fix the above issue, then I think it should be a separate ticket, because I think we should make this change across the codebase, in all places that we change a permission. 

In in each place a change happens, the permission would need to be changed for all versions of a release, or maintained on a single release (i.e. when creating an amendment, delete previous release's permissions after copying them to an amendment).

Reply:
Discussed this with Nick on Slack. I'll fix this bug here by applying. I've made a ticket to fix the wider issue: https://dfedigital.atlassian.net/browse/EES-2949